### PR TITLE
Layout refactor

### PR DESCRIPTION
**Pull request by mzient**
#### Why we need this PR?
- It adds a flexible and extandable layout to DALI tensors
- Refactoring to improve layout handling

#### What happened in this PR?
 - Explain solution of the problem, new feature added; what was changed, added, removed?
    * DALITensorLayout enum removed
    * TensorLayout used instead
    * All uses of layout enums replaced with strings
    * Switches over layout values replaced with queries and/or `layout.find(dim)`
 - What is most important part that reviewers should focus on?
 - Was this PR tested? How?
    * Old unit tests apply
 - Were docs and examples updated, if necessary?
    * Schemas updated where necessary

**JIRA TASK**: [DALI-1043]

### CODE REVIEW

**Comment by mzient** [comment]
!build

**Comment by mzient** [comment]
!build

**Comment by mzient** [comment]
!build

File: dali/pipeline/operators/op_schema.h [hunk 1]
```diff
@@ -207,12 +209,102 @@ class DLL_PUBLIC OpSchema {
...
+    }
+    input_layouts_[index] = layouts;
+    return *this;
+  }
+
+  /**
+   * @brief Sets input layout constraint and default for all inputs.
+   * @see InputLayout(int index, TensorLayout layout)
+   */
+  DLL_PUBLIC inline OpSchema& InputLayout(TensorLayout layout) {
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/pipeline/operators/op_schema.h]: How? It's already implemented in terms of another.

File: dali/pipeline/operators/op_schema.h [hunk 2]
```diff
@@ -207,12 +209,102 @@ class DLL_PUBLIC OpSchema {
...
+   * @brief Sets input layout constraints and default for given input.
+   *
+   * At run-time, when the operator encounters a tensor(list) with specified
+   * layout, but different than one provided to this function, error is raised.
+   *
+   * If the input tensor has no layout, the one provided to this function is assumed
+   * if number of dimensions matches. Otherswise, error is raised.
+   */
+  DLL_PUBLIC inline OpSchema& InputLayout(int index, TensorLayout layout) {
+    CheckInputIndex(index);
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/pipeline/operators/op_schema.h]: Right.

File: dali/pipeline/operators/op_schema.h [hunk 3]
```diff
@@ -207,12 +209,102 @@ class DLL_PUBLIC OpSchema {
...
+        if (l == layout)
+          return l;
+      std::stringstream ss;
+      ss << "The layout \"" << layout.c_str() << "\" does not match any of the allowed"
+        " layouts for input " << index << ". Valid layouts are:\n";
+      for (auto &l : input_layouts_[index])
+        ss << l.c_str() << "\n";
+      DALI_FAIL(ss.str());
+    }
+  }
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/pipeline/operators/op_schema.h]: It's not missing - the function ends with `throw`.

File: dali/pipeline/operators/resize/resize.cu [hunk 4]
```diff
@@ -61,6 +61,7 @@ void Resize<GPUBackend>::RunImpl(DeviceWorkspace &ws) {
   auto &output = ws.Output<GPUBackend>(0);
 
   RunGPU(output, input, ws.stream());
+  output.SetLayout(InputLayout(ws, 0));
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/pipeline/operators/resize/resize.cu]: RunGPU doesn't look at layouts.

File: dali/pipeline/operators/op_schema.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (mzient) [reply dali/pipeline/operators/op_schema.h]: Done.

**Comment by mzient** [comment]
!build

File: dali/pipeline/operators/crop/crop.cc [hunk 5]
```diff
@@ -38,12 +38,12 @@ template <>
 void Crop<CPUBackend>::DataDependentSetup(SampleWorkspace &ws) {
   const auto &input = ws.Input<CPUBackend>(0);
 
-  const DALITensorLayout in_layout = input.GetLayout();
-  DALI_ENFORCE(in_layout == DALI_NHWC || in_layout == DALI_NCHW
-            || in_layout == DALI_NFHWC || in_layout == DALI_NFCHW
-            || in_layout == DALI_NDHWC || in_layout == DALI_NCDHW,
+  const TensorLayout in_layout = InputLayout(ws, 0);
+  DALI_ENFORCE(in_layout.ndim() == input.shape().sample_dim());
+  DALI_ENFORCE(ImageLayoutInfo::HasChannel(in_layout) &&
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/pipeline/operators/crop/crop.cc]: `IsImage` / `IsVideo` does not require a channel dimension.

**Comment by mzient** [comment]
!build

**Comment by mzient** [comment]
!build

**Comment by mzient** [comment]
!build

File: include/dali/core/python_util.h [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+template <typename T>
+if_iterable<T> python_repr(std::ostream &os, const T &collection) {
+  os << "[";
+  bool first = true;
+  for (const auto &v : collection) {
+    if (!first)
+      os << ", ";
+    first = false;
```
Comment (jantonguirao) [inline include/dali/core/python_util.h]: this can go inside the `if`

**Pull request merged**